🚀 Feature Request: Show error.documentation_url in wrangler

### Describe the solution

`documentation_url` is a new field in the ErrorData interface for Cloudflare's v4 API responses:
https://github.com/cloudflare/cloudflare-typescript/blob/main/src/resources/shared.ts#L251
https://developers.cloudflare.com/api/resources/$shared/
The purpose is to make it easier for clients to redirect users to relevant documentation when encountering 4xx errors, instead of embedding links in the error text itself.

Current wrangler output in response to API errors:
```
✘ [ERROR] A request to the Cloudflare API (/accounts/[redacted]/workers/scripts/foo-bar) failed.

  Binding 'AI' of type 'ai' requires a Worker written in ES module format. [code: 100329] 
```

Desired wrangler output:
```
✘ [ERROR] A request to the Cloudflare API (/accounts/[redacted]/workers/scripts/foo-bar) failed.

  Binding 'AI' of type 'ai' requires a Worker written in ES module format. [code: 100329] 

To learn more about this error, visit https://developers.cloudflare.com/workers/reference/migrate-to-module-workers/
```

--- a/.changeset/lovely-lemons-count.md
+++ b/.changeset/lovely-lemons-count.md
@@ -1,0 +1,5 @@
+---
+"wrangler": patch
+---
+
+fix: include documentation_url in API Errors if provided

--- a/packages/wrangler/src/__tests__/cfetch-utils.test.ts
+++ b/packages/wrangler/src/__tests__/cfetch-utils.test.ts
@@ -47,7 +47,7 @@ describe("throwFetchError", () => {
 	mockAccountId();
 	mockApiToken();
 	runInTempDir();
-	mockConsoleMethods();
+	const std = mockConsoleMethods();
 
 	it("should include api errors, messages and documentation_url in error", async () => {
 		msw.use(
@@ -96,6 +96,75 @@ describe("throwFetchError", () => {
 				},
 			],
 		});
+	});
+
+	it("nested", async () => {
+		msw.use(
+			http.get("*/user", () => {
+				return HttpResponse.json(
+					createFetchResult(
+						null,
+						false,
+						[
+							{
+								code: 10001,
+								message: "error one",
+								documentation_url: "https://example.com/1",
+								error_chain: [
+									{
+										code: 10002,
+										message: "error two",
+										error_chain: [
+											{
+												code: 10003,
+												message: "error three",
+												documentation_url: "https://example.com/3",
+												error_chain: [
+													{
+														code: 10004,
+														message: "error 4",
+														documentation_url: "https://example.com/4",
+													},
+												],
+											},
+										],
+									},
+								],
+							},
+						],
+						["message one", "message two"]
+					)
+				);
+			})
+		);
+		await expect(runWrangler("whoami")).rejects.toMatchInlineSnapshot(
+			`[APIError: A request to the Cloudflare API (/user) failed.]`
+		);
+
+		expect(std.out).toMatchInlineSnapshot(`
+			"Getting User settings...
+
+			[31mX [41;31m[[41;97mERROR[41;31m][0m [1mA request to the Cloudflare API (/user) failed.[0m
+
+			  error one [code: 10001]
+			  To learn more about this error, visit: [4mhttps://example.com/1[0m
+
+			  - error two [code: 10002]
+
+			    - error three [code: 10003]
+			      To learn more about this error, visit: [4mhttps://example.com/3[0m
+
+			      - error 4 [code: 10004]
+			        To learn more about this error, visit: [4mhttps://example.com/4[0m
+
+			  message one
+			  message two
+
+			  If you think this is a bug, please open an issue at:
+			  [4mhttps://github.com/cloudflare/workers-sdk/issues/new/choose[0m
+
+			"
+		`);
 	});
 
 	it("should include api errors without messages", async () => {

--- a/packages/wrangler/src/__tests__/cfetch-utils.test.ts
+++ b/packages/wrangler/src/__tests__/cfetch-utils.test.ts
@@ -49,7 +49,7 @@ describe("throwFetchError", () => {
 	runInTempDir();
 	mockConsoleMethods();
 
-	it("should include api errors and messages in error", async () => {
+	it("should include api errors, messages and documentation_url in error", async () => {
 		msw.use(
 			http.get("*/user", () => {
 				return HttpResponse.json(
@@ -57,8 +57,20 @@ describe("throwFetchError", () => {
 						null,
 						false,
 						[
-							{ code: 10001, message: "error one" },
-							{ code: 10002, message: "error two" },
+							{
+								code: 10001,
+								message: "error one",
+								documentation_url: "https://example.com/1",
+							},
+							{
+								code: 10002,
+								message: "error two",
+								documentation_url: "https://example.com/2",
+							},
+							{
+								code: 10003,
+								message: "error three",
+							},
 						],
 						["message one", "message two"]
 					)
@@ -68,8 +80,15 @@ describe("throwFetchError", () => {
 		await expect(runWrangler("whoami")).rejects.toMatchObject({
 			text: "A request to the Cloudflare API (/user) failed.",
 			notes: [
-				{ text: "error one [code: 10001]" },
-				{ text: "error two [code: 10002]" },
+				{
+					text: "error one [code: 10001]\nTo learn more about this error, visit: https://example.com/1",
+				},
+				{
+					text: "error two [code: 10002]\nTo learn more about this error, visit: https://example.com/2",
+				},
+				{
+					text: "error three [code: 10003]",
+				},
 				{ text: "message one" },
 				{ text: "message two" },
 				{
@@ -85,13 +104,25 @@ describe("throwFetchError", () => {
 				return HttpResponse.json({
 					result: null,
 					success: false,
-					errors: [{ code: 10000, message: "error" }],
+					errors: [
+						{
+							code: 10000,
+							message: "error",
+							documentation_url: "https://example.com/1",
+						},
+						{ code: 10001, message: "error 1" },
+					],
 				});
 			})
 		);
 		await expect(runWrangler("whoami")).rejects.toMatchObject({
 			text: "A request to the Cloudflare API (/user) failed.",
-			notes: [{ text: "error [code: 10000]" }],
+			notes: [
+				{
+					text: "error [code: 10000]\nTo learn more about this error, visit: https://example.com/1",
+				},
+				{ text: "error 1 [code: 10001]" },
+			],
 		});
 	});
 });

--- a/packages/wrangler/src/cfetch/errors.ts
+++ b/packages/wrangler/src/cfetch/errors.ts
@@ -2,6 +2,7 @@ import { ParseError } from "../parse";
 
 export interface FetchError {
 	code: number;
+	documentation_url?: string;
 	message: string;
 	error_chain?: FetchError[];
 }

--- a/packages/wrangler/src/cfetch/index.ts
+++ b/packages/wrangler/src/cfetch/index.ts
@@ -207,6 +207,9 @@ function renderError(err: FetchError, level = 0): string {
 			.join("\n") ?? "";
 	return (
 		(err.code ? `${err.message} [code: ${err.code}]` : err.message) +
+		(err.documentation_url
+			? `\nTo learn more about this error, visit: ${err.documentation_url}`
+			: "") +
 		chainedMessages
 	);
 }

--- a/packages/wrangler/src/cfetch/index.ts
+++ b/packages/wrangler/src/cfetch/index.ts
@@ -198,17 +198,18 @@ function hasCursor(result_info: unknown): result_info is { cursor: string } {
 }
 
 function renderError(err: FetchError, level = 0): string {
+	const indent = "  ".repeat(level);
 	const chainedMessages =
 		err.error_chain
 			?.map(
 				(chainedError) =>
-					`\n${"  ".repeat(level)}- ${renderError(chainedError, level + 1)}`
+					`\n\n${indent}- ${renderError(chainedError, level + 1)}`
 			)
 			.join("\n") ?? "";
 	return (
 		(err.code ? `${err.message} [code: ${err.code}]` : err.message) +
 		(err.documentation_url
-			? `\nTo learn more about this error, visit: ${err.documentation_url}`
+			? `\n${indent}To learn more about this error, visit: ${err.documentation_url}`
 			: "") +
 		chainedMessages
 	);

